src/storage/sql_table.cpp

```diff
@@ -86,8 +376,75 @@ ProjectionMap SqlTable::ProjectionMapForInitializer(const ProjectionInitializerT
+  //    For each column not present in the old version, change the column id to the sentinel value
+  //    VERSION_POINTER_COLUMN_ID
+  for (uint16_t i = 0; i < out_buffer->NumColumns(); i++) {
+    TERRIER_ASSERT(out_buffer->ColumnIds()[i] != VERSION_POINTER_COLUMN_ID,
+                   "Output buffer should not read the version pointer column.");
+    catalog::col_oid_t col_oid = curr_dt_version.inverse_column_map.at(out_buffer->ColumnIds()[i]);
+    if (old_dt_version.column_map.count(col_oid) > 0) {
+      out_buffer->ColumnIds()[i] = old_dt_version.column_map.at(col_oid);
+    } else {
+      // TODO(Yashwanth): consider renaming VERSION_POINTER_COLUMN_ID, since we're using it for more than just that now
+      out_buffer->ColumnIds()[i] = VERSION_POINTER_COLUMN_ID;
+    }
+  }
+}
+
+template void SqlTable::ModifyProjectionHeaderForVersion<ProjectedRow>(ProjectedRow *out_buffer,
+                                                                       const DataTableVersion &curr_dt_version,
+                                                                       const DataTableVersion &old_dt_version,
+                                                                       col_id_t *original_col_id_store) const;
+template void SqlTable::ModifyProjectionHeaderForVersion<ProjectedColumns>(ProjectedColumns *out_buffer,
+                                                                           const DataTableVersion &curr_dt_version,
+                                                                           const DataTableVersion &old_dt_version,
+                                                                           col_id_t *original_col_id_store) const;
+
+std::unordered_set<catalog::col_oid_t> SqlTable::GetMissingColumnOidsForVersion(
+    const ProjectionMap &pr_map, const DataTableVersion &old_dt_version) const {
+  // Calculate the col_oids of out_buffer which are missing from the old version of datatable
+  std::unordered_set<catalog::col_oid_t> missing_col_oids;
+  for (const auto &it : pr_map) {
+    missing_col_oids.emplace(it.first);
+  }
+  for (const auto &it : old_dt_version.column_map) {
+    // Remove the col_oid from the missing_col_oids
+    missing_col_oids.erase(it.first);
+  }
+  return missing_col_oids;
+}
+
+template <class RowType>
+void SqlTable::FillDefaultValue(RowType *out_buffer, const catalog::col_oid_t col_oid,
+                                const ProjectionMap &pr_map) const {
+  TERRIER_ASSERT(default_value_map_.Find(col_oid) != default_value_map_.CEnd(),
+                 "Every column in schema must exist in default_value_map");
+  auto pair = default_value_map_.Find(col_oid)->second;
+  auto default_value = pair.first;
+  auto attr_size = pair.second;
+  // TODO(Sai): If this becomes a performance bottleneck, we can move this logic to ModifyProjectionHeaderForVersion
+  storage::StorageUtil::CopyWithNullCheck(default_value, out_buffer, attr_size, pr_map.at(col_oid));
```

Comment:
`CopyWithNullCheck` only does a shallow copy and `VarlenEntry` has no reference counting. You may need careful reasoning about how this will interact with the GC.